src/RoslynInsertionTool/RoslynInsertionTool/VersionsUpdater.cs

```diff
@@ -169,24 +188,67 @@ private void UpdateAssemblyVersionsFile(string assemblyName, Version newVersion)
                     break;
             }
 
-            var i = content.IndexOf(variableName);
-            if (i < 0)
+            // the structure of the file is:
+            //   header: opening tag:    <#+
+            //           comments:           //
+            //   values:                     const string AssemblyNameVersion = "<version-number>";
+            //   footer: closing tag:    #>
+            var lines = content.Split('\n').Select(line => line.TrimEnd('\r')).ToList();
+            var finalLines = new List<string>();
+            bool IsVersionLine(string line) => line.TrimStart().StartsWith("const");
+            var headerLines = lines.TakeWhile(line => !IsVersionLine(line)).ToList();
+            var valueLines = lines.Skip(headerLines.Count).TakeWhile(IsVersionLine).ToList();
+            var footerLines = lines.Skip(headerLines.Count + valueLines.Count).ToList();
+
+            // find the variable or the appropriate insertion location
+            var constExpression = "const string";
+            string GetLineVariableName(string line)
             {
-                throw new InvalidDataException($"Definition of {variableName} not found in '{fullPath}'.");
+                var startIndex = line.IndexOf(constExpression) + constExpression.Length + 1;
+                var endIndex = line.IndexOf(' ', startIndex + 1);
+                var variableLength = endIndex - startIndex;
+                return line.Substring(startIndex, variableLength);
             }
+            var sortedValueLines = valueLines.OrderBy(GetLineVariableName).ToList();
 
-            var versionStart = content.IndexOf('"', i + variableName.Length) + 1;
-            var versionEnd = content.IndexOf('"', versionStart);
-            if (versionStart <= 0 || versionEnd < 0)
+            // rather than get fancy, linearly search through the sorted list and update or add as appropriate
+            // the file is small so this is fine
+            bool valueUpdated = false;
+            var newLine = $@"    const string {variableName} = ""{newVersion.ToFullVersion()}"";";
+            for (int lineIndex = 0; lineIndex < sortedValueLines.Count; lineIndex++)
             {
-                throw new InvalidDataException($"File '{fullPath}' doesn't have expected format.");
+                var line = sortedValueLines[lineIndex];
+                var currentVariableName = GetLineVariableName(line);
+                var comparison = String.Compare(currentVariableName, variableName, StringComparison.OrdinalIgnoreCase);
+                if (comparison == 0)
+                {
+                    // found exact match, replace this line
+                    var versionStart = line.IndexOf('"') + 1;
```

Comment:
BTW: all these .IndexOfs that will return -1 if not found are terrifying me. Wonder if we should have ThrowingIndexOf or something.

Reply:
Good call, I'll do that.